File: C++/Algorithms/Graph Theory Algorithms/kosaraju.cpp
#include <iostream>
#include <list>
#include <stack>
using namespace std;
class Graph{
    int V;
    list<int> *l;
    list<int>*lr;

    public:
        Graph(int V){
            this->V=V;
            l=new list<int>[V];
            lr=new list<int>[V];
        }
        void addEdge(int a,int b){
            l[a].push_back(b);
            lr[b].push_back(a);
        }
        void dfs(int node, bool visited[],stack<int>& s){
            visited[node]=true;
            for(auto nbr:l[node]){
                if(!visited[nbr]){
                    dfs(nbr,visited,s);
                }
            }
            s.push(node);
        }
        void dfs_reverse(int node,bool visited[]){
            visited[node]=true;
            cout<<node<<" ";
            for(auto nbr:lr[node]){
                if(!visited[nbr]){
                    dfs_reverse(nbr,visited);
                }
            }
        }
        void kosaraju(){
            bool* visited=new bool[V];
            for(int i=0;i<V;i++){
                visited[i]=false;
            }
            stack<int> s;
            for(int i=0;i<V;i++){
                if(!visited[i]){
                    dfs(i,visited,s);
                }
            }
            //use lr for reversed graph
            for(int i=0;i<V;i++){
                visited[i]=false;
            }
            while(!s.empty()){
                int t=s.top();
                if(visited[t]==false){
                    cout<<"Elements in a component are: ";
                    dfs_reverse(t,visited);
                    cout<<endl;
                }
                s.pop();
            }
        }
};
int main(){
    Graph g(10);
    g.addEdge(0,1);
    g.addEdge(1,2);
    g.addEdge(2,3);
    g.addEdge(3,0);
    g.addEdge(2,4);
    g.addEdge(4,5);
    g.addEdge(5,6);
    g.addEdge(6,4);
    g.addEdge(7,6);
    g.addEdge(7,8);
    g.addEdge(8,7);
    g.addEdge(8,9);
    g.kosaraju();
}

File: Java/Algorithms/Graphs/kosaRaju.java

import java.util.ArrayList;
import java.util.Stack;
//A directed graph is strongly connected if there is a path between all pairs of vertices. 
//A strongly connected component (SCC) of a directed graph is a maximal strongly connected subgraph.

// A class to find strongly connected component (SCC) in directed graph
public class kosaraju {
	 // 3 steps to apply kosaraju
	 // 1) find the topological sort ( i will create another file for it )
	 // 2) then take transpose of the graph
	 // 3) then just apply dfs in topological sort
	 void kosaRaju(ArrayList<ArrayList<Integer>> adj, int n)
	    {
	        int vis[] = new int[n]; 
	        Stack<Integer> st = new Stack<Integer>();  // stack is used to store the topological sort ,and we will apaly dfs accordingly
	        for(int i = 0;i<n;i++) {
	        	if(vis[i] == 0) { // if already visited ,don't visit it
	        		dfs(i, st, adj, vis); 
	        	}
	        }

	        ArrayList<ArrayList<Integer> > transpose = new ArrayList<ArrayList<Integer> >(); // to store the transpose of graph
			
			for (int i = 0; i < n; i++)  
				transpose.add(new ArrayList<Integer>());

			for(int i = 0;i<n;i++) {
				vis[i] = 0; 
				for(Integer it: adj.get(i)) {
					transpose.get(it).add(i); 
				}
			}
			
            // step 3 -- apply dfs in topo sort
			while(st.size() > 0) {
				int node = st.peek(); 
				st.pop(); 
				if(vis[node] == 0) {
					System.out.print("SCC: "); 
					revDfs(node, transpose, vis);
					System.out.println(); 
				}
			}

	    }
	 // step 1 -- to find topo sort
	private void dfs(int node, Stack<Integer> st, ArrayList<ArrayList<Integer>> adj, int vis[]) {
		vis[node] = 1;
		for(Integer it : adj.get(node)) {
			if(vis[it] == 0) {
				dfs(it, st, adj, vis); 
			}
		}
        // when dfs complete we will store in stack
		st.push(node); 
	}
    // apply dfs
	private void revDfs(int node, ArrayList<ArrayList<Integer>> transpose, int vis[]) {
		vis[node] = 1;
		System.out.print(node + " "); 
		for(Integer it : transpose.get(node)) {
			if(vis[it] == 0) {
				revDfs(it, transpose, vis); 
			}
		}
	}

   
 
}
